stdlib/src/builtin/string.mojo

```diff
@@ -1979,6 +1980,94 @@ struct String(
             )
         return String(buf^)
 
+    @always_inline
+    fn format[*Ts: Stringable](self, *args: *Ts) raises -> String:
+        """Format a template with *args.
+
+        Example of manual indexing:
+
+        ```mojo
+        print(
+            String("{0} {1} {0}").format(
+                "Mojo", 1.125
+            )
+        ) #Mojo 1.125 Mojo
+        ```
+
+        Example of automatic indexing:
+
+        ```mojo
+        var x = String("{} {}").format(
+            True, "hello world"
+        )
+        print(x) #True hello world
+        ```
+
+        ⚠️ Raising functions does not work in the parameter domain.
+
+        Args:
+            args: Values passed as variadic arguments.
+
+        Parameters:
+            Ts: The types of the variadic arguments.
+
+        Returns:
+            The formated template.
+
+        Variadic arguments are required to implement `Stringable`.
+
+        """
+        var entries = _FormatCurlyEntry.create_entries(self)
+        var manual_indexing_count = 0
+        var automatic_indexing_count = 0
+        for e in entries:
+            if e[].value.isa[Int]():
+                manual_indexing_count += 1
+                if e[].value[Int] >= len(args):
+                    raise ("Index " + str(e[].value[Int]) + " not in *args")
+            elif e[].value.isa[NoneType]():
+                automatic_indexing_count += 1
+            elif e[].value.isa[String]():
+                raise "Index " + e[].value[String] + " not in kwargs"
+            if manual_indexing_count and automatic_indexing_count:
+                raise "Cannot both use manual and automatic indexing"
+
+        if automatic_indexing_count > len(args):
+            raise ("Automatic indexing require more args in *args")
+
+        var res: String = ""
```

Comment:
followup idea: It would be nice if we could preallocate the correct result size here. Let's keep this patch simple, but in a followup you could probably modify `create_entries` to count the number of characters that are are not in the fields, right?

Reply:
Yes thought about it too, something like `capacity = len(self) - len(args) ?`
Let's keep this review open so we don't forget about it ?
That way we can do optimization at the end ?